src/Select.purs

```diff
@@ -90,266 +77,300 @@ derive instance ordVisibility :: Ord Visibility
 -- | about searches when time has expired).
 data InputType = Text | Toggle
 
--- | The component state
-type State st =
+type SelectInput m =
   { inputType :: InputType
-  , search :: String
-  , debounceTime :: Milliseconds
-  , debounceRef :: Maybe (Ref (Maybe Debouncer))
+  , search :: Maybe String
+  , debounceTime :: Maybe Milliseconds
+  , getItemCount :: HookM m Int
+  , pushNewSearch  :: String -> HookM m Unit
+  , pushVisibilityChanged :: Visibility -> HookM m Unit
+  , pushSelectedIdxChanged :: Int -> HookM m Unit
+  }
+
+type SelectState =
+  { search :: String
   , visibility :: Visibility
   , highlightedIndex :: Maybe Int
-  , getItemCount :: {| st } -> Int
-  | st
   }
 
-type Debouncer =
-  { var :: AVar Unit
-  , fiber :: Fiber Unit
-  }
+newtype SelectReturn m = SelectReturn
+  { search :: String
+  , visibility :: Visibility
+  , highlightedIndex :: Maybe Int
 
-type Input st =
-  { inputType :: InputType
-  , search :: Maybe String
-  , debounceTime :: Maybe Milliseconds
-  , getItemCount :: {| st } -> Int
-  | st
+  , setFocus :: Boolean -> HookM m Unit
+  , setVisibility :: Visibility -> HookM m Unit
+  , clearSearch :: HookM m Unit
+
+  , toggleProps
+      :: forall toggleProps
+       . Array (HP.IProp (ToggleProps toggleProps) (HookM m Unit))
+  , itemProps
+      :: forall itemProps
+       . Int
+      -> Array (HP.IProp (ItemProps itemProps) (HookM m Unit))
+  , containerProps
+      :: forall containerProps
+       . Array (HP.IProp (onMouseDown :: ME.MouseEvent | containerProps) (HookM m Unit))
+  , inputProps
+      :: forall inputProps
+       . Array (HP.IProp (InputProps inputProps) (HookM m Unit))
```

Comment:
These were originally functions `Array _ -> Array _` so that `append` could be pre-applied. That was done because we want to make sure these props get overridden by the user's props, if they supply a duplicate. It's also a little more convenient at the call site -- `props <> [ ]` or `append props []` vs. `setProps []`.

What do you think of returning to that approach? Did you have a reason to move to this one instead?

Reply:
Personally, I thought `props` is better than `setProps []`. If a user wants to override them, wouldn't `props <> [ override1Prop ]` work, too?

Reply:
It does work, it’s just that it’s really easy to write it the other way around and not know why your override isn’t working

Reply:
Ah.... that makes sense. And if you, for whatever reason, decided to combine multiple libraries' props together, you'd get
```purescript
-- no overrides
setSelectProps (setFormlesProps (setCustomHookProps []))

-- with overrides
setSelectProps (setFormlesProps (setCustomHookProps 
  [ overrideSelectProp
  , overrideFormlessProp
  , overrideCustomHookProp
  ]))
```
and even the order in which you 'set' the props matters because Formless could override Select's corresponding prop.

I'll make the change back to `setProps`. If we want to approach this in some other manner, we can in a future PR.

Reply:
Done in latest commit.